### PR TITLE
SNOW-1526511: unhardcode top level domain

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-hx**
teamwork issue 1046

### CODE REVIEW

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: lib/connection.c [hunk 1]
```diff
@@ -594,34 +593,29 @@ char * STDCALL encode_url(CURL *curl,
     const char *amp = "&";
     size_t amp_size = strlen(amp);
 
+    if (host_empty)
+    {
+      SET_SNOWFLAKE_ERROR(error, SF_STATUS_ERROR_BAD_CONNECTION_PARAMS,
+                          "Invalid host trying to create encoded url",
+                          SF_SQLSTATE_UNABLE_TO_CONNECT);
+      goto cleanup;
+    }
```
Comment (sfc-gh-dstempniak) [inline lib/connection.c:602]: What about 6th point?
```
detected top level domain should be logged with a respective message at the INFO level:
Connecting to GLOBAL Snowflake domain
or
Connecting to CHINA Snowflake domain
respectively.
```

Comment (sfc-gh-ext-simba-hx) [reply lib/connection.c:602]: oops I missed that. Will add.

Comment (sfc-gh-ext-simba-hx) [reply lib/connection.c:602]: added

File: deps/curl/lib/vtls/sf_ocsp.c [hunk 2]
```diff
@@ -2244,10 +2271,33 @@ void initOCSPCacheServer(struct Curl_easy *data)
...
+    else
+    {
+      // It's basically impossible not finding top domain in host.
+      // Use "com" as default just in case.
+      top_domain = "com";
+    }
+
+    if (strcasecmp(top_domain, "com") == 0)
+    {
+      infof(data, "Connecting to GLOBAL Snowflake domain");
```
Comment (sfc-gh-dstempniak) [inline deps/curl/lib/vtls/sf_ocsp.c:2288]: What about private links? 
If `ocsp_cache_server_url_env != NULL` and `host` is specified there will be no info message at all, am I right?

File: lib/client.c [hunk 3]
```diff
@@ -432,11 +437,23 @@ _snowflake_check_connection_parameters(SF_CONNECT *sf) {
...
+                sf_sprintf(buf, sizeof(buf), "%s.%s.snowflakecomputing.cn",
+                         sf->account, sf->region);
+                log_info("Connecting to CHINA Snowflake domain");
+            }
+            else
+            {
+                sf_sprintf(buf, sizeof(buf), "%s.%s.snowflakecomputing.com",
+                         sf->account, sf->region);
+                log_info("Connecting to GLOBAL Snowflake domain");
+            }
```
Comment (sfc-gh-dstempniak) [inline lib/client.c:452]: Maybe we could move this check outside `if (!sf->host) {..}` condition and always check postfix of `sf->host`?
In such a case we will have only one `if .. else ...` condition and there will be no need to log similar information in `sf_ocsp.c`.

File: deps/curl/lib/vtls/sf_ocsp.c (hunk 2 above)
Comment (sfc-gh-dprzybysz) [reply deps/curl/lib/vtls/sf_ocsp.c:2288]: We have a separate issue for private links support SNOW-1538117

File: lib/client.c (hunk 3 above)
Comment (sfc-gh-ext-simba-hx) [reply lib/client.c:450]: good idea. done.

File: lib/client.c [hunk 4]
```diff
@@ -432,14 +438,45 @@ _snowflake_check_connection_parameters(SF_CONNECT *sf) {
...
+        top_domain++;
+    }
+    else
+    {
+        // It's basically impossible not finding top domain in host.
+        // Log the entire host just in case.
+        top_domain = sf->host;
+    }
+
+    if (strcasecmp(top_domain, "com") == 0)
```
Comment (sfc-gh-dprzybysz) [inline lib/client.c:471]: please change the log to expose CHINA and GLOBAL only like in JDBC https://github.com/snowflakedb/snowflake-jdbc/blob/60e9e541794bbd80a33b00cf8039c8512d8c5361/src/main/java/net/snowflake/client/core/SFSession.java#L625-L627

**Review by sfc-gh-dprzybysz — CHANGES_REQUESTED** [review]

File: lib/client.c (hunk 4 above)
Comment (sfc-gh-ext-simba-hx) [reply lib/client.c]: done

**Review by sfc-gh-dprzybysz — APPROVED** [review]

**Review by sfc-gh-dstempniak — APPROVED** [review]

**Pull request merged**